lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/VectorShortDecimalFixedWidthByteArrayBatchDecoder.java

```diff
@@ -0,0 +1,371 @@
+            implements ShortDecimalDecoder
+    {
+        private static final VectorShuffle<Byte> SHUFFLE = VectorShuffle.fromArray(
+                ByteVector.SPECIES_256,
+                new int[] {
+                        7, 6, 5, 4, 3, 2, 1, 0,
+                        15, 14, 13, 12, 11, 10, 9, 8,
+                        23, 22, 21, 20, 19, 18, 17, 16,
+                        31, 30, 29, 28, 27, 26, 25, 24
+                },
+                0);
+
+        @Override
+        public void decode(SimpleSliceInputStream input, long[] values, int offset, int length)
+        {
+            byte[] inputArr = input.getByteArray();
+            int inputOffset = input.getByteArrayOffset();
+            int inputBytesRead = 0;
+            int outputOffset = offset;
+            while (length >= 4) {
+                ByteVector.fromArray(ByteVector.SPECIES_256, inputArr, inputOffset + inputBytesRead)
+                        .rearrange(SHUFFLE)
+                        .reinterpretAsLongs()
+                        .intoArray(values, outputOffset);
+                inputBytesRead += 32;
+                length -= 4;
+                outputOffset += 4;
+            }
+            input.skip(inputBytesRead);
+
+            // Decode the remaining values "normally" as it would read data out of bounds
+            while (length > 0) {
+                values[outputOffset++] = Long.reverseBytes(input.readLongUnsafe());
+                length--;
+            }
+        }
+    }
+
+    private static final class BigEndianReader7
+            implements ShortDecimalDecoder
+    {
+        private static final VectorShuffle<Byte> SHUFFLE = VectorShuffle.fromArray(
+                ByteVector.SPECIES_256,
+                new int[] {
+                        7, 6, 5, 4, 3, 2, 1, 0,
+                        14, 13, 12, 11, 10, 9, 8, 7,
+                        21, 20, 19, 18, 17, 16, 15, 14,
+                        28, 27, 26, 25, 24, 23, 22, 21
+                },
+                0);
+
+        @Override
+        public void decode(SimpleSliceInputStream input, long[] values, int offset, int length)
+        {
+            byte[] inputArr = input.getByteArray();
+            int inputOffset = input.getByteArrayOffset();
+            int inputBytesRead = 0;
+            int outputOffset = offset;
+            while (length > 4) {
+                // We read redundant bytes and then ignore them. Sign bit is propagated by `>>` operator
+                ByteVector.fromArray(ByteVector.SPECIES_256, inputArr, inputOffset + inputBytesRead)
+                        .rearrange(SHUFFLE)
+                        .reinterpretAsLongs()
+                        .lanewise(VectorOperators.ASHR, 8)
+                        .intoArray(values, outputOffset);
+                inputBytesRead += 28;
+                length -= 4;
+                outputOffset += 4;
+            }
+            // Decode the last 5 values "normally" as it would read data out of bounds
```

Comment:
Shouldn't this comment say "the last 4 values"? length = 5 will be handled by the loop above.

Reply:
Also, it may be cleaner check remaining bytes in the loop condition instead and ensure there are at least 32 bytes available to read. Then, the reader of this code doesn't need to do mental math to figure out how many values are required to cover 32 bytes worth of input.

You would need to calculate the number of bytes required to decode `length` worth of values up front for the loop condition.